_posts/2024-10-03-pixi-basics.md

```diff
@@ -0,0 +1,358 @@
+
+Benchmark 1: pixi add pangolin && pixi install && pixi remove pangolin
+
+  Time (mean ± σ):     160.734 s ±  5.329 s    [User: 41.234 s, System: 105.102 s]
+  Range (min … max):   154.892 s … 167.771 s    4 runs
+ 
+Benchmark 2: conda create -y -p condaenv pangolin && rm -rf condaenv
+
+  Time (mean ± σ):     321.362 s ±  5.470 s    [User: 31.665 s, System: 39.065 s]
+  Range (min … max):   315.114 s … 328.384 s    4 runs
+ 
+Summary
+  'pixi add pangolin && pixi install && pixi remove pangolin' ran
+    2.00 ± 0.07 times faster than 'conda create -y -p condaenv pangolin && rm -rf condaenv'
+Benchmark 1: pixi add ivar && pixi install && pixi remove ivar
+
+  Time (mean ± σ):     24.782 s ±  0.354 s    [User: 7.629 s, System: 7.915 s]
+  Range (min … max):   24.505 s … 25.275 s    4 runs
+ 
+Benchmark 2: conda create -y -p condaenv ivar && rm -rf condaenv
+
+  Time (mean ± σ):     37.005 s ±  0.452 s    [User: 6.882 s, System: 5.673 s]
+  Range (min … max):   36.575 s … 37.606 s    4 runs
+ 
+Summary
+  'pixi add ivar && pixi install && pixi remove ivar' ran
+    1.49 ± 0.03 times faster than 'conda create -y -p condaenv ivar && rm -rf condaenv'
+
+```
+
+</details>
+
+For the complicated installation, pixi is 2.21x faster than conda.
+For individual installations, it is consistently more than 1.4x faster than conda.
+
+## Day to day operations
+
+Why would you even go through all this?
+If you have read this post up to here, I am guessing that you already use conda and do not appreciate being told to fix what isn't broken.
+
+Instead, consider that there could be times when it is way more appropriate to use pixi.
+For example, let's say that you are working in a project folder where you make your own "lab notebook" with markdown.
+The project is simple; it is reads mapped to a reference genome.
+
+What were your methods? Your lab notebook/readme says that you downloaded `SRR123` using `fastq-dump`, downloaded `NC_123`, and then mapped the reads with `bwa`. Finally, you sorted and indexed your bam file with `samtools`.
+
+In most of our notebooks, we might have forgotten to specify what software we used or maybe what versions.
+The most careful of us might have had a conda environment file.
+The 0.01% of us might have even been careful enough to save the environment into the actual project folder.
+
+However, with pixi, the environment creation is fast enough that you don't have to save it forever in your project folder.
+The initialization is folder/project based too -- it is simple enough to write into your notebook to start with `pixi shell` and to have the `pixi.toml` file as part of your project.
+This workflow gets much better if you are in a collaborative environment with multiple hands on the same folder.
+
+That is all to say that I don't think that there is a definitive reason to switch to pixi -- I think it just flows better.
+The same way that you start off a development project with `git init`.
```

Comment:
Does it need a "Caveats" section?

If so I've got one:

- If installing python via pixi, note that you need to install pip separately before using it.

Reply:
I'm not sure if that specific caveat is exactly true but I have addressed what I found in the documentation from this page <https://pixi.sh/latest/advanced/pyproject_toml/#initial-setup-of-the-pyprojecttoml-file> with

## Caveats

There are many caveats to switching over from conda.
In fact, there is a [whole article](https://pixi.sh/latest/switching_from/conda/) about it.
Here are some noteworthy one caveats.

* Python itself should be added to the dependencies section, e.g.,

```toml
[dependencies]
python = ">=3.9"
```

* pip requirements can be added under the `[project]` section, e.g.,

```toml
[project]
dependencies = [
    "numpy",
    "pandas",
    "matplotlib",
]
```

You can specify the required python version with `requires-python` without including it in the specific env like so:

```toml
[project]
requires-python = ">=3.9"
```

Reply:
Looks great!